### PR TITLE
udev: introduce udevadm control --show-children-max

### PR DESCRIPTION
**Pull request by msekletar**

### CODE REVIEW

**Comment by msekletar** [comment]
This PR was motivated by downstream RFE bug. It gets the job done, but code is not generic at all. udevadm <-> udevd communication was so far one way, i.e. udevd sent no data back to the client. This newly introduced option is an exception. Should we introduce some more generic infrastructure in udev to allow for cleaner implementation of bidirectional communication? I don't know...

**Comment by msekletar** [comment]
@martinpitt any idea why CI failed?

**Comment by martinpitt** [comment]
@msekletar : I think yes. This got in the middle of updating the packaging to 231 and I forgot to update the symbols files until this morning. Retried.

File: man/udevadm.xml [hunk 1]
```diff
@@ -440,6 +440,14 @@
           </listitem>
         </varlistentry>
         <varlistentry>
+          <term><option>-c</option></term>
+          <term><option>--show-children-max=</option></term>
```
Comment (mkmkme) [inline man/udevadm.xml]: Do you actually need "=" here? Looks like this option does not take any arguments.

Comment (msekletar) [inline man/udevadm.xml]: Nope, that is copy-paste error. Thanks!

**Comment by poettering** [comment]
What's the usecase for this precisely?

**Comment by msekletar** [comment]
Some RHEL users were encountering udev (with RHEL6) related issues on big iron machines. IIRC, they reported various hangs and timeouts and solution was to adjust number of events dispatched at the same time. But they also wanted to examine current value. That is not possible at the moment.

If above rationale is not sufficient I can get back to them and ask about details.

**Comment by evverx** [comment]
> But they also wanted to examine current value. That is not possible at the moment.

Why not to use `systemtap` or so?

``` sh
$ sudo stap -v -e 'probe process("/usr/lib/systemd/systemd-udevd").{function("on_ctrl_msg"),function("on_ctrl_msg").return} { printf("%s %u\n", is_return() ? "new" : "old", @var("arg_children_max")) }'
...
# run udevadm control -m 100
old 12
new 100

# run udevadm control -m 50
old 100
new 50
^C
```

**Comment by msekletar** [comment]
Sure, systemtap or gdb would work. However, using either of these is not an ideal solution for production environments.

**Comment by poettering** [comment]
To me it appears that this should just be something we log about, no? Why do we need a proper API for this?

We could even attach a message id to such a log message so that people can retrieve the right value directly via "journalctl -b MESSAGE_ID=..."...

**Comment by msekletar** [comment]
That was my initial approach because implementation is nicer, i.e. doesn't change current model where udev sends no data back to the client. But in downstream bugzilla they said that they don't find such solution useful. And honestly, I don't blame them. OTOH, I know that this PR is bit controversial.

**Comment by poettering** [comment]
@msekletar why wouldn't that be "useful"?

If I type "journalctl -b MESSAGE_ID=..." or "udevadm control --show-children-max", where's the difference?

**Comment by msekletar** [comment]
I'd find it useful. But I guess that in their production environments they don't have logs stored locally. Most likely they are getting transferred to remote log aggregator via legacy syslog protocol, i.e. no MESSAGE_ID available.

Honestly, I this particular case it doesn't help me to introduce mechanism upstream that I can't backport downstream to RHEL6 too. If you feel uneasy about this PR then we can drop it. I am fine with that too.

**Comment by poettering** [comment]
i do feel uneasy about it. I also don't buy the argument that logs aren't good enough for that, I must say.

What about this: "systemctl status" shows a status string from each service that supplies one via sd_notify(). udev currently just sets it to "Processing...". Maybe we should improve that, and change this to "Processing, with %u children at max". I think that would be generally useful, easy to query and requires no new interfaces or protocols anywhere. Does that sounds acceptable? This would essentially be a single-line change, trivially easy to backport...

**Comment by msekletar** [comment]
I like your idea about sd_notify(). I'll prep patch for that. Thanks!

**Comment by evverx** [comment]
> This would essentially be a single-line change

`udevadm control -m N` should change the status message too. So, this is not a single-line change.

> "systemctl status" shows a status string from each service that supplies one via sd_notify().

`sd_notify` can block `systemd-udevd`. Shouldn't we use a non-blocking notification instead?

**Comment by msekletar** [comment]
> sd_notify can block systemd-udevd. Shouldn't we use a non-blocking notification instead?

Hmm, I did not realize this. Now I recall that some time ago there was the same issue in journald and non trivial rework was necessary to use non-blocking notifications.

**Comment by poettering** [comment]
Well, we call sd_notify() anyway, and it's generally OK for services spawned by PID 1 to block on PID 1, as long as we don't block from PID 1 on that service. For udev we really don't, hence this is OK.

Generally, it's PID 1 that provides services to everything else, hence it's generally fine for everything else to block on PID 1, but as soon as we have the opposite direction in there too, we have a problem.

hence: using sd_notify() for this should be perfectly fine, and we use it already, hence it's really just an issue of slightly changing the string we send, not a change of whether we send it or not.

**Comment by evverx** [comment]
> we call sd_notify() anyway, and it's generally OK for services spawned by PID 1 to block on PID 1, as long as we don't block from PID 1 on that service. For udev we really don't, hence this is OK.

Well, there is no a deadlock. This is OK.
But I don't really understand why  should `udevadm control -m new-child-max` potentially block `systemd-udevd` on `PID1`

**Comment by poettering** [comment]
> But I don't really understand why should udevadm control -m new-child-max potentially block systemd-udevd on PID1

Not sure I grok the question? sd_notify() only blocks if the socket is contended, which it normally isn't. It normally is asynchronous in behaviour.

**Comment by evverx** [comment]
> It normally is asynchronous in behaviour

Normally yes.

I mean this scenario:
-  A long-running `daemon-reload` (https://github.com/systemd/systemd/issues/3241, https://github.com/systemd/systemd/issues/3353 or so)
- Other services call `sd-notify`, nobody reads, the queue is full (this is possible: some SElinux policies don't allow `bump_unix_max_dgram_qlen` for example)
- Run `udevadm control -m new-child-max`
- `systemd-udevd` changes the `arg_child_max`, calls `sd_notify` and blocks on  `PID1` until the end of the `daemon-reload`

**Comment by poettering** [comment]
@evverx sure, but there's nothing wrong with blocking in that case, or is there? I mean, it's completely OK if udev blocks in that case, in particular as that should only be an exceptional case...

**Comment by evverx** [comment]
> it's completely OK if udev blocks in that case

Well, we can sleep, say, 1m30s. Is this OK?

The original issue:

> they reported various hangs and timeouts

We will introduce another potential delay. And I don't understand why :)

**Comment by poettering** [comment]
@evverx well, the sd_notify() STATUS= thing only exists so that services can pass a quick status string to the service manager. That's a highly useful functionality, that lots of services already make use for. Should there be real-life issues with the queue limits then we should work on fixing those, and make things more scalable (but afaics they are only theoretical at this point?), but the key here really is that the interface is good and useful, and udev should just make use of it like everything else. In particular as it already makes use of it, except that its strings so far weren't overly informative.

**Comment by poettering** [comment]
@msekletar any update on this one?

**Comment by msekletar** [comment]
-> #4036

**Comment by msekletar** [comment]
Sorry for the delay.

**Pull request closed without merging**